File: 1137/main.py
from collections import deque

class Vertex:
    def __init__(self, n=0):
        self.out = []
        self.n = n

def explore(u, ans):
    ans.append(u)
    q = deque()
    v = u
    while v.out:
        w = v.out.pop()
        v = w
        q.append(v)
    while q:
        explore(q.popleft(), ans)

def main():
    v = [Vertex(i) for i in range(10001)]
    
    n = int(input())
    last = tot = 0
    
    for _ in range(n):
        m_and_rest = list(map(int, input().split()))
        m = m_and_rest[0]
        last = m_and_rest[1]
        for u in m_and_rest[2:]:
            v[u].n = u
            v[last].out.append(v[u])
            last = u

    ans = []
    explore(v[last], ans)
    print(len(ans) - 1, end=' ')
    for vertex in ans:
        print(vertex.n, end=' ')

if __name__ == "__main__":
    main()

File: 1137/main.cpp
#include <stdio.h>
#include <vector>
#include <queue>
 
struct vertex{
    std::vector<vertex*> out;
    int n;
};
 
void explore(vertex* u, std::vector<vertex*>& ans){
    ans.push_back(u);
    std::queue<vertex*> q;
    vertex* v = u;
    while(!v->out.empty()){  
        auto w = v->out.back();
        v->out.pop_back();
        v = w;
        q.push(v);
    }
    while(!q.empty()){ 
         explore(q.front(), ans);
         q.pop();
    }
 }
 
int main(){
    int n;
    std::vector<vertex> v(10001);
 
    scanf("%d", &n);
    int m, last, tot = 0;
 
    for(int i = 0; i < n; i++){
        scanf("%d", &m);
        scanf("%d", &last);
        for(int j = 1; j <= m; j++){
            int u;
            scanf("%d", &u);
            v[u].n = u;
            v[last].out.push_back(&v[u]);
            last = u;
        }
    }
    
    std::vector<vertex*> ans;
    explore(&v[last], ans);
    printf("%d ", ans.size()-1);
    for(auto it = ans.begin(); it < ans.end(); it++)
        printf("%d ", (*it)->n);
    return 0;
 }